### PR TITLE
Fixed tofu test when module has no resource

### LINKED ISSUE #1359: `tofu test` - If module has no resources, its outputs are unusable
### OpenTofu Version

```shell
1.6.2
```

### OpenTofu Configuration Files

`main.tf`:
```hcl
module "first" {
  source = "./first"
}

resource "tfcoremock_simple_resource" "resource" {}
```

`first/main.tf`:
```hcl
module "second" {
  source = "../second"
}

output "id" {
  value = module.second.id
}
```

`second/main.tf`:
```hcl
resource "tfcoremock_simple_resource" "resource" {}

output "id" {
  value = tfcoremock_simple_resource.resource.id
}
```

`main.tftest.hcl`:
```hcl
run "run" {
  command = apply

  assert {
    condition     = module.first.id != 0
    error_message = "Fail"
  }
}
```

### Debug Output

.

### Expected Behavior

Test should pass

### Actual Behavior

Test fails:

```
│ Error: Unknown condition run
│ 
│   on main.tftest.hcl line 5, in run "run1":
│    5:     condition     = module.first.id != 0
│ 
│ Condition expression could not be evaluated at this time.
```

### Steps to Reproduce

1. `tofu init`
2. `tofu test`

### Additional Context

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by siddharthasonker95**
Objective:
Fix for tofu test - When module has no resources, its outputs are unusable and the test fails throwing below error:

> │ Error: Unknown condition run
│ 
│   on main.tftest.hcl line 5, in run "run1":
│    5:     condition     = module.first.id != 0
│ 
│ Condition expression could not be evaluated at this time.

Changes:

- Added new method `checkPlannedState()` in `test_context.go` which compares the planned state against the current state and updates the current state with any missing modules from the planned state. It returns a SyncState representing the updated state.
- Added relevant test cases in `test_test.go`

Resolves #1359 

## Target Release

1.7.0

### CODE REVIEW

**Comment by Yantrio** [comment]
@siddharthasonker95 is this ready for review?

File: internal/tofu/test_context.go [hunk 1]
```diff
@@ -67,6 +67,11 @@ func (ctx *TestContext) EvaluateAgainstPlan(run *moduletest.Run) {
 }
 
 func (ctx *TestContext) evaluate(state *states.SyncState, changes *plans.ChangesSync, run *moduletest.Run, operation walkOperation) {
+	if ctx.Plan != nil && ctx.Plan.PlannedState != nil &&
+		len(ctx.State.Modules) != len(ctx.Plan.PlannedState.Modules) {
+		state = checkPlannedState(ctx.State, ctx.Plan.PlannedState)
+	}
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply internal/tofu/test_context.go:73]: The current state does not include the module that has no resources, which is why [GetModule()](https://github.com/opentofu/opentofu/blob/63c88507a878513cf7f860eceefd024a6da78b57/internal/tofu/evaluate.go#L394) is unable to evaluate the output variable and throws error.
I will add a comment here to explain the idea behind this logic.

File: internal/tofu/test_context.go [hunk 2]
```diff
@@ -186,3 +193,17 @@ func (ctx *TestContext) evaluate(state *states.SyncState, changes *plans.Changes
...
+//
+// If a module has no resources, it is included in the current state to ensure that its output variables are usable.
+func synchronizeStates(state, plannedState *states.State) *states.SyncState {
+	newState := state.DeepCopy()
+	for key, value := range plannedState.Modules {
+		if _, exists := newState.Modules[key]; !exists {
+			newState.Modules[key] = value
+		}
+	}
+	return newState.SyncWrapper()
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply internal/tofu/test_context.go:208]: The [state](https://github.com/opentofu/opentofu/blob/7abee8c4f1cc3980b530e19891cba8d42067a501/internal/tofu/test_context.go#L69) does not include the module that has no resource due to which it's output variables are unusable. So for it's output to be usable that module with no resource should be added to the state.

I believe deep copying the state won't discard the sync state information as we are deep copying the state present in the context (i.e. [ctx.State](https://github.com/opentofu/opentofu/blob/7abee8c4f1cc3980b530e19891cba8d42067a501/internal/tofu/test_context.go#L74)) not the syncState(in the [parameter](https://github.com/opentofu/opentofu/blob/7abee8c4f1cc3980b530e19891cba8d42067a501/internal/tofu/test_context.go#L69)). These state have been [synchronised](https://github.com/opentofu/opentofu/blob/7abee8c4f1cc3980b530e19891cba8d42067a501/internal/tofu/test_context.go#L59) just before the [evaluate()](https://github.com/opentofu/opentofu/blob/7abee8c4f1cc3980b530e19891cba8d42067a501/internal/tofu/test_context.go#L69) function is triggered. I can update the state before sync as well if the current changes may encounter concurrency issues.

[the remark this answers is not in the document]
Comment (Yantrio) [reply internal/tofu/test_context.go:208]: that works for me. I misread the state that was being passed in. I thought it was the syncstate from the parameter and not the ctx.State 👍 that's fine,=

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +8,7 @@ ENHANCEMENTS:
 
 BUG FIXES:
 * Added a check in the `tofu test` to validate that the names of test run blocks do not contain spaces. ([#1489](https://github.com/opentofu/opentofu/pull/1489))
+* Fix for `tofu test` when the module has no resources, its outputs become unusable. ([#1409](https://github.com/opentofu/opentofu/pull/1409))
```
Comment (Yantrio) [inline CHANGELOG.md:11]: Can you talk about the positive effect of this now? maybe saying something like...

`tofu test` now supports accessing module outputs when the module has no resources.

what do you think? I think we should always try and focus on the positive (the fix) and not the negative (the bug)

Comment (siddharthasonker95) [reply CHANGELOG.md]: sure, that makes more sense.

**Review by Yantrio — APPROVED** [review]
I'm happy with this if @janosdebugs is too :)

**Pull request merged**